Objects/methodobject.c

```diff
@@ -160,7 +160,10 @@ PyCMethod_GetClass(PyObject *op)
 static void
 meth_dealloc(PyCFunctionObject *m)
 {
-    _PyObject_GC_UNTRACK(m);
+    // The Py_TRASHCAN mechanism requires that we be able to
+    // call PyObject_GC_UnTrack twice on an object.
```

Comment:
I don't understand that, but thanks for the comment!

Reply:
@tim-one added this comment in 803526b9e222077c5f9391d7e7348156eb7ba7da but the original change is by @nascheme, committed by @gvanrossum in ff413af605d7aab610272c5668bc870ebcbf56ef with the following description:
```
    This is Neil's fix for SF bug 535905 (Evil Trashcan and GC interaction).

    The fix makes it possible to call PyObject_GC_UnTrack() more than once
    on the same object, and then move the PyObject_GC_UnTrack() call to
    *before* the trashcan code is invoked.
```

Reply:
If the comment is unclear, maybe we should try to improve it.  I believe the main point is that, since the object uses the trashcan, the function version of untrack is used rather than the macro version.  The macro is not safe to call twice because it assumes the object is a part of the GC double linked list.  At least, that's my memory of it.

Why the trashcan calls untrack more than once is obscure and I don't recall why it does so.  I would need to study the code.  My guess would be because it (ab)uses the GC head pointers to store the object in a special trashcan linked list.

Reply:
Maybe it would be more clear if you would say that the code before `Py_TRASHCAN_BEGIN` can be executed more than once.  Calling the macro version of untrack twice is not safe while the function version is safe.